accepted/00000-jsonb-pillar-db-storage.md

```diff
@@ -0,0 +1,225 @@
+  port: 5432
+  user: spacewalk
+
+ext_pillar:
+  - postgres:
+      - query: "SELECT pillar from suseSaltPillars WHERE target = '*'"
+      - query: "SELECT pillar from suseSaltPillars WHERE target = (SELECT CONCAT('org_id:', org_id::VARCHAR) FROM rhnServer AS S LEFT JOIN suseMinionInfo AS M on S.id = M.server_id WHERE M.minion_id = %s)"
+      - query: "SELECT pillar from suseSaltPillars WHERE target IN (SELECT CONCAT('group_id:', server_group_id::VARCHAR) FROM rhnServerGroupMembers AS G LEFT JOIN suseMinionInfo AS M ON G.server_id = M.server_id WHERE M.minion_id = %s)"
+      - query: "SELECT pillar from suseSaltPillars WHERE target = %s"
+```
+
+This example contains three different queries:
+- Universal pillars looking for target `*`
+- Organization pillars looking for target `org_id:<organization id>`
+- Group pillars. This is done by looking up salt client `server_id` using its `minion_id` joining server group membership table to obtain `server_group_id` set where salt client is a member. Then looking up target `group_id:<group id>` in this result set.
+- Minion pillars looking for `minion_id`
+
+Salt automatically merge results from these queries, latest with the highest priority.
+
+### Upstream changes needed
+
+salt.pillar.postgres module, resp. salt.pillar.sql_base is not yet designed to work with JSON results directly from database and expects always to get results in *key:value* format. To enable JSON output, small patch need to be negotiated with upstream (patch against pre-blackened version):
+
+```diff
+diff --git a/salt/pillar/sql_base.py b/salt/pillar/sql_base.py
+index eeb2269f77..dae0b36204 100644
+--- a/salt/pillar/sql_base.py
++++ b/salt/pillar/sql_base.py
+@@ -352,8 +352,13 @@ class SqlBaseExtPillar(six.with_metaclass(abc.ABCMeta, object)):
+                         # dict, descend.
+                         crd = crd[ret[i]]
+
++            # We have just one field without any key, assume returned row is already a dict
++            # aka JSON storage
++            if self.num_fields == 1:
++                crd.update(ret[0])
++
+             # If this test is true, the penultimate field is the key
+-            if self.depth == self.num_fields - 1:
++            elif self.depth == self.num_fields - 1:
+                 nk = self.num_fields-2  # Aka, self.depth-1
+                 # Should we and will we have a list at the end?
+                 if ((self.as_list and (ret[nk] in crd)) or
+```
+
+Tracked in [PR#59777](https://github.com/saltstack/salt/pull/59777).
+
+## Maintaining consistency
+
+Responsibility to provide up to date data remains, as is the case today, on the component changing the data. That means, when information about minion is changed, then relevant pillar generator must be called to refresh the pillar in database.
+However with the pillar in database and loosely (no foreign key relation) connected to minions and groups, database trigger which can remove pillar entries when minion or group is removed:
```

Comment:
What advantages and disadvantages you see with using triggers rather than enhancing Java code to maintain consistency when, e.g., a minion is removed from a group?

Reply:
I understand and agree about the strong consistency guarantee argument, although I am also reminded of this short story:

http://thecodelesscode.com/case/42

which is the reason I normally prefer code-level enforcing (where refactorings are typically eased by IDEs).

That said, I fully trust you on having the last word and ultimately decide on this aspect.